syntax = "proto3";

package chat;

option go_package = "/;api";

message ChatID {
    string chat_id=1;
}

message UserChats {
    repeated ChatID chat_ids=1;
}

message UserID {
    string user_id=1;
}

enum MessageType {
    Normal=0;
}

message MessageInfo {
    string message_json=1;
}

message Message {
    MessageInfo message=1;
    MessageType type=2;
}

message ChatInformation {
    repeated Message messages=1;
    string title=2;
    repeated UserID participants=3;
}

message ParticipantInfo {
    string name=1;
    string nickName=2;
    UserID user_id=3;
}

enum Direction {
    chanAfter=0;
    chanBefore=1;
}

message MessageChan {
    string message_id=1;
    Direction direction=2;
}

message Messages {
    repeated Message messages=1;
}

service Chat {
    rpc getUserChats(UserID) returns(UserChats);
    rpc getChatInfo(ChatID) returns(ChatInformation);
    rpc getParticipantInfo(UserID) returns(ParticipantInfo);
    rpc getMessages(MessageChan) returns(Messages);

    rpc MessageExchange(stream Message) returns(stream Message);
}
